File: Linear_Control_Course/Chapter19/Lesson1/Chapter19_Lesson1.cpp
#include <iostream>
#include <complex>
#include <vector>

struct LeadCompensator {
    double Kc;
    double alpha;
    double T;

    std::complex<double> eval(const std::complex<double>&s) const {
        std::complex<double> num = 1.0 + alpha * T * s;
        std::complex<double> den = 1.0 + T * s;
        return Kc * num / den;
    }
};

struct Plant {
    // Example: G(s) = 1 / (s (s + 2))
    std::complex<double> eval(const std::complex<double>&s) const {
        return 1.0 / (s * (s + 2.0));
    }
};

int main() {
    LeadCompensator C{1.0, 4.0, 0.1}; // example parameters
    Plant G;

    std::vector<double> omega_list{0.5, 1.0, 2.0, 5.0};

    for (double omega : omega_list) {
        std::complex<double> s(0.0, omega);
        std::complex<double> L = C.eval(s) * G.eval(s);

        double mag = std::abs(L);
        double phase_rad = std::arg(L);
        double phase_deg = phase_rad * 180.0 / M_PI;

        std::cout << "omega = " << omega
                  << ", |L(j omega)| = " << mag
                  << ", phase(L) = " << phase_deg << " deg\n";
    }

    return 0;
}

File: Tutorials/Codes/Linear_Control_Course/Chapter19/Lesson1/Chapter19_Lesson1.java
import org.apache.commons.math3.complex.Complex;

public class LeadCompensator {
    private final double Kc;
    private final double alpha;
    private final double T;

    public LeadCompensator(double Kc, double alpha, double T) {
        this.Kc = Kc;
        this.alpha = alpha;
        this.T = T;
    }

    public Complex eval(Complex s) {
        Complex num = Complex.ONE.add(s.multiply(alpha * T));
        Complex den = Complex.ONE.add(s.multiply(T));
        return num.divide(den).multiply(Kc);
    }

    public static Complex plant(Complex s) {
        // Example plant G(s) = 1 / (s (s + 2))
        return Complex.ONE.divide(s.multiply(s.add(2.0)));
    }

    public static void main(String[] args) {
        LeadCompensator C = new LeadCompensator(1.0, 4.0, 0.1);
        double[] omegaList = {0.5, 1.0, 2.0, 5.0};

        for (double omega : omegaList) {
            Complex s = new Complex(0.0, omega);
            Complex L = C.eval(s).multiply(plant(s));

            double mag = L.abs();
            double phaseRad = L.getArgument();
            double phaseDeg = Math.toDegrees(phaseRad);

            System.out.println("omega = " + omega
                + ", |L(j omega)| = " + mag
                + ", phase(L) = " + phaseDeg + " deg");
        }
    }
}
